### PR TITLE
fig-align cases for feature-format matrix and supporting tooling

### PR DESCRIPTION
**Pull request by gordonwoodhull**
Here are 10 test cases for `fig-align` across four formats

image

There are four areas, so we may not want to merge this draft PR this at once:
1. The tests and manually verified `quality` results. Some automatic testing for typst; I couldn't get the latex tests to work, so I've commented those out. 
2. `detailed` parameter on the feature-format-matrix, which reads all `*.qmd` and adds another level to the trie
3. Partial fix for #8767 - Typst currently scores best in the matrix 🥳 except for HTML where `fig-align` always works. 
4. A proof-of-concept `render-all-formats.py` - not sure if this is the right language or path for this tool, just something to start discussion. This is invaluable for testing because I can e.g. `open results/*/docx/*.docx` to flip through all the docx case results to assess quality. It also tries to save intermediate artifacts and it is the start of being able to deploy feature-format matrix artifacts.

### Testing notes
* Note that these tests introduce *expected failures*. I propose that the `quality` result gate the test result: if `quality <= 0` then a failing test is okay. Possibly succeeding a test where `quality <= 0` is also an error. I haven't looked at how to implement this.
* Currently I need to use `keep-typ: true` and not `output-ext: typ` in order to have tests that I can also render to PDF via Typst. I gather this is a bug and should be handled the same way as latex, which does not need `output-ext: tex`? Haven't dug into it yet.
* Would be nice to record test results across the matrix, but we don't currently record test results at that granularity. I haven't looked at the code. May be possible by scraping logs.
* I tried including tests with `{{< include tests.qmd >}}` but I guess metadata merging doesn't work here (and I haven't looked at the code to find out why). This would be nice for maintenance of tests where the same test can be used for many cases. (I'm currently using the same tests for all Typst cases, but it's not that simple in general.)

### Feature-format matrix UI notes
* Would be nice to move `document.qmd` to the top of the cases, but I haven't looked at how to do this.
* It's annoying that the first column is so narrow, looked briefly at Tabulator docs but couldn't figure out how to change this. Might be nice to add the resize columns module, but I haven't looked at how to add ES modules yet.

### CODE REVIEW

**Comment by cscheid** [comment]
> Note that these tests introduce expected failures.

I think this is OK, as long as we don't run the tests in the feature-format matrix in CI. Our policy is that we consider a CI test-suite failure to be breakage that needs to be fixed ASAP. 

> I propose that the quality result gate the test result: if quality <= 0 then a failing test is okay. Possibly succeeding a test where quality <= 0 is also an error.

This would require a certain amount of coordination between the YAML in `_quarto: tests` and the feature-format YAML in the documents themselves. The test verifiers in the `.qmd` files are driven by the way `smoke-all.test.ts` runs. I like the idea of tying those two things together, but the implementation would be a bit messy. Maybe what it means is that we need to (eventually) have an explicit `test-feature-format-matrix.test.ts` file that understands the feature-format YAML and can connect the tests to the quality field.

> Would be nice to record test results across the matrix, but we don't currently record test results at that granularity. I haven't looked at the code. May be possible by scraping logs.

All of our tests eventually go through `Deno.test` and their infrastructure. It's likely that they have some record-keeping mechanism, but I haven't looked at all into it. As we lean more into our test suite, it would be nice for us to be able to look at the results of a run in CI in a better way than scrolling up and down the logs.

**Comment by cscheid** [comment]
> I tried including tests with {{< include tests.qmd >}} but I guess metadata merging doesn't work here (and I haven't looked at the code to find out why).

This is another Quarto bug; we should have an API that lets document metadata be accessed "after the merging", as if it were in Pandoc. Right now, access is (effectively) ad hoc. My work on single-file projects and some of the new functions there has, in principle, unlocked our ability to create this API. Let me look into it.

**Review by cscheid — APPROVED** [review]
This looks good to me.

(We're not at all consistent about the following rule, but) In the long run, I think we would (minorly!) prefer our general scripts to be Deno scripts rather than Python scripts. The `.qmd` file I created used `python` code cells so that we could run it in Jupyter,  but Deno has the advantage that we _know_ it's installed in any Quarto deployment; technically speaking, it's possible for people to have Quarto installed without Python or Jupyter available.

It's not worth our time for you to rewrite this one in Deno, but just as a future note, if something is equally easy to do in typescript vs other languages, leaning into .ts is marginally preferred.

**Comment by gordonwoodhull** [comment]
Thanks @cscheid! Good points. Yes I'll translate the `render-all-formats.py` script to Deno / Typescript sometime - this was just a quick way for me to get started. 

> All of our tests eventually go through Deno.test and their infrastructure. It's likely that they have some record-keeping mechanism, but I haven't looked at all into it. 

I tried [enabling `--junit-path`](https://docs.deno.com/runtime/manual/basics/testing/#reporters) but it just produced xml with our stdout embedded in it. I think we would need to provide structured results somehow. I'll look into it, as I'm sure I will be doing a lot of work with test matrices this quarter. 

> I like the idea of tying those two things together, but the implementation would be a bit messy. Maybe what it means is that we need to (eventually) have an explicit test-feature-format-matrix.test.ts file that understands the feature-format YAML and can connect the tests to the quality field.

This sounds good. Again, I expect to be doing a lot of this for each missing Typst feature, so I can look into this too.

In the meantime, I'll see if @cderv has any concerns and merge in the next day or two. Let's talk about any related ideas on Thursday.

**Comment by gordonwoodhull** [comment]
Interestingly, the cases that don't work for Typst are precisely the ones which use `#figure()`

We currently don't try to align Typst figures and according to typst/typst#1409 it's not possible without some sort of override. The (horrible) workaround presented on that issue does work. 👿 

```typst
#show figure: it => align(right)[
  #it.body  
  #v(10pt, weak: true)
  #emph[#it.supplement #it.counter.display(it.numbering)]: #it.caption 
]
```

**Comment by cderv** [comment]
> Currently I need to use keep-typ: true and not output-ext: typ in order to have tests that I can also render to PDF via Typst. I gather this is a bug and should be handled the same way as latex, which does not need output-ext: tex? Haven't dug into it yet.

Regarding this, this is known, and `output-ext: tex` will also not run the PDF. If you look into our test, we usually don't test the PDF rendering. We do use 
````yaml
    latex:
      ensureFileRegexMatches:
````

So we convert to `latex` and test the content of the output file which will be `tex`. 

For some doc, ware also testing pdf rendering by adding `pdf` to the render format
````yaml
_quarto:
  tests:
    pdf: null
    latex:
      ensureFileRegexMatches:
        - ['The first paragraph\.[\s\S]+\\endgraf[\s\S]+The second paragraph']
        - []
````

In other place, we only test the correct pdf rendering with some resources.
````yaml
_quarto:
  tests:
    pdf:
      fileExists: 
        outputPath: simple-svg.pdf
````
In that case no need to keeping the `.tex` file.


So we should probably have an equivalent for `ensureTypstFileRegexMatches()` for when we want to test the PDF rendering, something like `ensurePdfLaTeXFileRegexMatches`. This would work the same where `keep-tex: true` would be required. 

So if I sumup currently how this works  for typst

- `format: typst` will render to `.typ` then convert to `.pdf` - we need `keep-typ: true` for now, and use `ensureTypstFileRegexMatches` in that case. 
- `format: typst` + `output-ext: typ` will just convert to `.typ` with no PDF rendering - in that case `ensureFileRegexMatches` is to be used. 

- `format: pdf` renders to pdf and `keep-tex: true` will keep the `.tex` file but we don't have a `ensurePdfLaTeXFileRegexMatches` for testing the regex in the intermediate file.
- `format: latex` exists in Pandoc - this should be equivalent in Quarto to `format: pdf` + `output-ext: tex`. We use `format: latex` when we want to test content in intermediate with `ensureFileRegexMatches`

All this is related to 
* https://github.com/quarto-dev/quarto-cli/issues/7039

Hope this helps clarify - this definitely needs improvement. 

I'll review the rest of the PR and comments to see if I have other feedback.

**Comment by cderv** [comment]
> I tried including tests with {{< include tests.qmd >}} but I guess metadata merging doesn't work here (and I haven't looked at the code to find out why)

I think it does not because we do read metadata in `smoke-all.test.ts` the input file before any resolution of `include` shortcode 
https://github.com/quarto-dev/quarto-cli/blob/265e3dc002455fc1e3e56fe2e2bbf731e5ef0055/tests/smoke/smoke-all.test.ts#L182-L191
https://github.com/quarto-dev/quarto-cli/blob/265e3dc002455fc1e3e56fe2e2bbf731e5ef0055/tests/smoke/smoke-all.test.ts#L74-L76

So `_quarto` needs to be in the metadata in the input file for now. This would require some change - probably something to consider when we'll be trying (at some point maybe ?) to allow `_quarto` to work with a standard `quarto render` - I don't think it does right now. It only works with `smoke-all.test.ts` AFAIK

File: dev-docs/feature-format-matrix/qmd-files/layout/fig-align/toplevel-id-caption.qmd [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  tests:
+# couldn't get either of these to work
+#    pdf:
+#      ensureFileRegexMatches:
+#      -
+#        - "\\\\hfill\\\\includegraphics\\{100.png\\}"
+#        - '\hfill\includegraphics\{100.png\}'
+#      - []
+    typst:
+      ensureTypstFileRegexMatches:
```
Comment (cderv) [inline dev-docs/feature-format-matrix/qmd-files/layout/fig-align/toplevel-id-caption.qmd:26]: As I explained in my other answer you need to use `latex`
```suggestion
    latex:
      ensureFileRegexMatches:
      -
        - "\\\\hfill\\\\includegraphics\\{100.png\\}"
        - '\hfill\includegraphics\{100.png\}'
      - []
    typst:
      ensureTypstFileRegexMatches:
```

You can add `pdf: null` also if you want to test pdf rendering with LaTeX, and not just the output

File: dev-docs/feature-format-matrix/qmd-files/layout/fig-align/toplevel-id-nocaption.qmd [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  tests:
+# couldn't get either of these to work
+#   pdf:
+#     ensureFileRegexMatches:
+#     -
+#       - "\\\\hfill\\\\captionsetup\\{labelsep=none\\}\\\\includegraphics\\{100.png\\}"
+#       - '\\hfill\\captionsetup\{labelsep=none\}\\includegraphics\{100.png\}'
+#     - []
+    typst:
+      ensureTypstFileRegexMatches:
```
Comment (cderv) [inline dev-docs/feature-format-matrix/qmd-files/layout/fig-align/toplevel-id-nocaption.qmd:26]: Same here
```suggestion
    latex:
      ensureFileRegexMatches:
      -
        - "\\\\hfill\\\\captionsetup\\{labelsep=none\\}\\\\includegraphics\\{100.png\\}"
        - '\\hfill\\captionsetup\{labelsep=none\}\\includegraphics\{100.png\}'
      - []
    typst:
      ensureTypstFileRegexMatches:
```

**Review by cderv — APPROVED** [review]
I think I am missing something about how to run the feature format matrix. or it is not working correctly for me on Windows. 

Some questions: 
- Why not a `code-r-label.qmd` to go with the `code-python-label.qmd` ? 
- Should we have also a `code-markdown-*.qmd` variable (using `engine: markdown)` using the Divs syntax in markdown only ? It should work too I believe - [Figure Divs](https://quarto.org/docs/authoring/figures.html#figure-divs)

Not related to this PR though - but it would be good to have README in `dev-docs\feature-format-matrix` or  a specific doc in `dev-docs\` about this so that we document how to render the page and documents. 

Other global feedback on feature matrix: 

we should probably have a at least a `requirement.txt` or better a `pipenv` configuration for this folder so that we can be sure to track the dependencies for Python. Using **renv** will be also better (or just  `DESCRIPTION`) to manage R environment for this feature matrix project.  It could also be common with our test environment (which is used when run with `./run-tests.sh '../dev-docs/feature-format-matrix/qmd-files/**/*.qmd'`)

Anyhow, LGTM - it was more general thought sharing; Thanks a lot for this PR @gordonwoodhull ! 

Regarding my rendering problem, I'll check if this is windows support problem or not, but it shouldn't block this PR as it definitely works for you.

**Comment by gordonwoodhull** [comment]
Thanks @cerv! I fixed the latex tests and will add the other cases you mentioned.

Yes, I keep getting tripped up by the format latex pdf typst stuff. I will get used to it eventually, but I imagine this is pretty confusing for power users as well. Thanks for the link to the epic.

**Comment by gordonwoodhull** [comment]
I'm going to merge this if the tests succeed, since I'm working on other things for a while.

The script is now in Deno Typescript and currently runs all documents in parallel (each document serial for formats).

**Pull request merged**